text/0000-array-private-classes.md

```diff
@@ -0,0 +1,366 @@
+they iterate.
+
+# How We Teach This
+
+This change keeps the code compatible in the vast majority of cases. When client
+classes are defining objects of these now private types, the reason is usually
+to get access to the function `rewind` that was not defined in `Iterator`. By
+adding the interface `CollectionIterator`, client code can easily be adapted,
+replacing `ArrayValues[A]` by `CollectionIterator[A]`.
+
+Also, client code generally uses these functions to iterate on the returned types
+and does not try to access the iterator directly but is interested by the iterated
+items. When client code refers to the iterator type, that's generally useless and
+the code can be rewritten to be made shorter and more future proof.
+
+A [search on Github Pony code](https://github.com/search?q=%22ArrayValues%22+language%3APony&type=code)
+finds 24 files using the class `ArrayValues`, of which 6 are copies of `array.pony` file.
+
+For instance, in
+[xml2xpath.pony](https://github.com/redvers/pony-libxml2/blob/bbca5d98d48854bfec2c6ee110220873ecc4df34/pony-libxml2/xml2xpath.pony#L41),
+the code can be changed from
+
+```pony
+  fun values(): ArrayValues[Xml2node, this->Array[Xml2node]]^ ? =>
+    if (allocated) then
+      ArrayValues[Xml2node, this->Array[Xml2node]](nodearray)
+    else
+      error
+    end
+```
+
+to
+
+```pony
+  fun values(): CollectionIterator[Xml2node]^ ? =>
+    if (allocated) then
+      nodearray.values()
+    else
+      error
+    end
+```
+
+In this sample, the developer was not really concerned by the type of the iterator
+but that the `values` function must return an `CollectionIterator` over `Xml2node`.
+The new version makes the code simpler to understand.
+
+This change in `array.pony` and other collections will break such code but it
+can be easily adapted to use the new API. And it will make the standard library
+easier to learn by reducing the number of public types.
+
+# How We Test This
+
+Pony tests must continue to pass. No additional tests are need as after review
+the existing coverage in Pony standard library tests is sufficient
+
+# Drawbacks
+
+Will break any existing code that uses any of the classes that are currently
+public and will be made private by this RFC.
+
+# Alternatives
+
+1. Stay as is. Continue the
+[discussion on Zulip](https://ponylang.zulipchat.com/#narrow/stream/189959-RFCs/topic/Make.20Array.20iterators.20private).
+
+2. Update the existing concrete classes to include `rewind` where needed.
+
+3. Instead of defining a new type of iterator with the `CollectionIterator` interface,
+we can consider only the rewindable part of it in an `Rewindable` interface:
+
+```pony
+interface Rewindable[A]
+  fun ref rewind(): A^
+    """
+    Rewind the type `A`.
+    """
+```
+
+Then one can create a rewindable iterator by combining these two interfaces into
+a intersection type:
+
+```pony
+type CollectionIterator[A] is (Iterator[A] & Rewindable[Iterator[A]])
+```
```

Comment:
This is my preference, aka how I would write my code, but I have had conversations which swayed me against this. I do not, as I sit here now, recall the reasoning that swayed me, but I am leaving a review comment here to voice my position on this alternative.